File: top_100_liked/python3/course_schedule.py
class Solution:
    def __init__(self):
        self.flag = 1
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        adj = [[] for _ in range(numCourses)]
        visited = [0] * numCourses
        for a, b in prerequisites:
            adj[a].append(b)
        for n in range(numCourses):
            if(not visited[n]):
                self.dfs(n, adj, visited)
        return self.flag
    
    # 0 is unvisited, 1 is in process, 2 is done
    def dfs(self, i, adj, visited):
        visited[i] = 1 
        for j in adj[i]:
            if(not visited[j]):
                self.dfs(j, adj, visited)
            elif(visited[j] == 1):
                self.flag = 0
        visited[i] = 2

File: top_100_liked/cpp/course_schedule.cpp
class Solution {
public:
    // dfs cycle detection
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        vector<vector<int> > edges(numCourses);
        // all unvisited initially
        vector<int> visited(numCourses, 0);
        for(auto e : prerequisites){
            edges[e[0]].push_back(e[1]);
        }
        
        for(int i = 0; i < numCourses; i++){
            if(visited[i] == 0)
                if(dfs(i, edges, visited)) 
                    return false;
        }
        
        return true;
    }
    // return true if there is a cycle
    bool dfs(int node, vector<vector<int> >& edges, vector<int>& visited){
        visited[node] = 1;
        for(auto n : edges[node]){
            if(visited[n] == 0)
                if(dfs(n, edges, visited))
                    return true;
            if(visited[n] == 1)
                return true;
            // don't do anything if visited[n] == 2
        }
        visited[node] = 2;
        
        return false;
    }
};
